### PR TITLE
Sites Dashboard: Persist launch status state in URLs

### LINKED ISSUE #65323: Sites Management Dashboard: Manage filter state through query parameters
### What

It should be possible to open the sites list (`/sites-dashboard`) in a certain state.

### Why

This would allow for a better user experience: imagine scenarios where people bookmark the sites list link at a certain state and they want to quickly apply the same filters.

Another very common scenario is to share a link with the filters so other people could see the same page as the sender.

### How

The easiest and most practical way would be to parse query parameters on page load while updating them on every filter update in the UI.

### Example usage

The idea that came up in https://github.com/Automattic/wp-calypso/pull/65296#discussion_r915008706 makes it possible to:

1. open the sites list;
2. change filters;
3. grab the URI in the address bar;
4. open the same link elsewhere...

And see the same filters being applied to the sites list.

### Acceptance criteria

- [ ] The app reads query parameters and applies the filters to the site list on load
- [ ] Changing a filter in the sites list updates the query parameters

### PR DESCRIPTION
**Pull request by danielbachhuber**
Fixes #65323

## Proposed Changes

Persists the launch status tab filter state in URLs (e.g. `/sites-dashboard?status=launched` and `/sites-dashboard?status=private`):

image

'All' sites remains as `/sites-dashboard`.

One thing worth noting: `TabPanel` renders `<button>`, so we have to call `page()` to update the URL when a tab item is selected. I _think_ this is a horrible hack that's bad for accessibility, but I'm not 100% confident. I don't think it should be a PR blocker, though.

## Testing Instructions

1. Navigate to `/sites-dashboard`.
2. Click on the 'Launched' tab.
3. Observe the URL as `/sites-dashboard?status=launched`, and only Launched sites listed.
4. Click on the 'Private' tab.
5. Observe the URL as `/sites-dashboard?status=private`, and only Private sites listed.
6. Click on the 'Coming soon' tab.
7. Observe the URL as `/sites-dashboard?status=coming-soon`, and only Coming soon sites listed.
8. Refresh the page.
9. Observe the URL as `/sites-dashboard?status=coming-soon`, and only Coming soon sites listed.
10. Manually navigate to `/sites-dashboard?status=foobar`.
11. Observe the 'All' tab selected with all sites visible.

### CODE REVIEW

**Comment by zaguiini** [comment]
I like this! I like the idea behind it, and the implementation is very clean.

The only thing I don't like is that we're using part of the path to indicate the current tab, when this is actually state... I believe it would be better if we used query parameters to store/update the tab to prevent page collisions in the future (e.g. if we create a `/private` page).

[Considering where we're at and our next steps](https://github.com/Automattic/wp-calypso/issues/65323), it makes even more sense to use a query parameter. This PR would introduce the mechanism for loading/updating the query parameters into SMD, while other PRs could extend it to work with other filters (i.e. the search bar).

The code looks good and it's working correctly. I'd like some feedback on this design decision, though -- let me know what you think!

**Comment by danielbachhuber** [comment]
> I like this! I like the idea behind it, and the implementation is very clean.

@zaguiini Thanks!



> The only thing I don't like is that we're using part of the path to indicate the current tab, when this is actually state... I believe it would be better if we used query parameters to store/update the tab to prevent page collisions in the future (e.g. if we create a `/private` page).
> 
> [Considering where we're at and our next steps](https://github.com/Automattic/wp-calypso/issues/65323), it makes even more sense to use a query parameter. This PR would introduce the mechanism for loading/updating the query parameters into SMD, while other PRs could extend it to work with other filters (i.e. the search bar).
> 
> The code looks good and it's working correctly. I'd like some feedback on this design decision, though -- let me know what you think!

Fair enough. Personally, I was split on whether the tabs represented separate pages or filtering state. I agree with your perspective, though.

I've switched to a `status` query arg for now in 5d46514f3c092c572c629fc9de4e737ce7ef432c

File: client/sites-dashboard/index.ts [hunk 1]
```diff
@@ -9,5 +9,5 @@ export default function () {
 		return;
 	}
 
-	page( '/sites-dashboard', sitesDashboard, makeLayout, clientRender );
+	page( '/sites-dashboard/:launchStatus?', sitesDashboard, makeLayout, clientRender );
```
Comment (zaguiini) [inline client/sites-dashboard/index.ts]: I think we need to revert this change.

Comment (danielbachhuber) [reply client/sites-dashboard/index.ts]:
```suggestion
	page( '/sites-dashboard', sitesDashboard, makeLayout, clientRender );
```

Sure!

In the future, you're welcome to make this change yourself.

**Review by vykes-mac — APPROVED** [review]
I agree with @zaguiini , nice clean design. I've tested it and it works as expected. 👍

**Comment by wojtekn** [comment]
@danielbachhuber I haven't got a chance to look at this one yet, but it's worth checking it with https://github.com/Automattic/wp-calypso/pull/65462 I'm working on.

**Comment by danielbachhuber** [comment]
> I haven't got a chance to look at this one yet, but it's worth checking it with #65462 I'm working on.

@wojtekn My PR doesn't seem to have an impact on yours. Is there something specific you're concerned about?

**Comment by wojtekn** [comment]
> @wojtekn My PR doesn't seem to have an impact on yours. Is there something specific you're concerned about?
 
My PR hooks the Sites Dashboard into `/home/` route as a replacement for the current sites selector, so the dashboard won't be available under `/sites-dashboard` anymore. 

I will check how those works together, it may be that it will work well OOTB. Just wanted to make you aware of those changes.

**Comment by danielbachhuber** [comment]
> My PR hooks the Sites Dashboard into `/home/` route as a replacement for the current sites selector, so the dashboard won't be available under `/sites-dashboard` anymore.
> 
> I will check how those works together, it may be that it will work well OOTB. Just wanted to make you aware of those changes.

@wojtekn Makes sense, thanks for the heads up!

It seems like the URL construction logic might need some form of abstraction if the Sites Dashboard might be embedded in multiple URLs.
